### ISSUE TITLE
protect TaskWorker from main loop crashes

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#5216]
in
https://github.com/dmwm/CRABServer/blob/master/src/python/TaskWorker/MasterWorker.py
should al least wrap with a try so that if something bad happens (corrupt task info in DB e.g.) some more info is printed and possibly TW close down nicely. Better if various failures are individually trapped. E.g. missing/invalid/inconsistent task information. Better yet to skip and log the bad task and go on processing the others.

See this example:
https://cms-logbook.cern.ch/elog/Analysis+Operations/899

**Comment by mmascher** [comment]
I agree, it's not nice that there was a failure in the MasterWorker, but no message was printed in /data/srv/TaskManager/current/logs/processes/proc.master.txt , and instead it went to nohup.out

**Comment by mmascher** [comment]
I spare you all the details that I started writing here[1] but I'll write the conclusion anyhow.

I saw there were two `crab kill` executed in30 seconds and I was fearing a concurrency problem, but apparently this is more related to the fact that in the status we relabeled a `QUEUED` state to `KILLED` https://github.com/dmwm/CRABServer/blob/master/src/python/CRABInterface/HTCondorDataWorkflow.py#L574-L581 . That opens the door to a concurrency problem since two crab kill are executed in parallel and overwrites the command field.

There are six operations we do for a kill (I concentrated on database queries that could have changed the status or have an impact on future queries that changed the status):

C1: do a crab status (SELECT)
C2: if the status of the task is ok (e.g.: not QUEUED or HOLDING), then put status=NEW and command=KILL
TW1: put on hold N tasks that are NEW
TW2: get the data related to those tasks
TW3: for each task put it to QUEUED and inject it
TW4: when the kill is done put status=KILLED and command=NONE

This is how chronologically the two kill were handled:

1.C1 17:58:53
1.C2 17:58:55
1.TW1 (timestamp not relevant) 
1.TW2
1.TW3 17:59:08
2.C1 17:59:28
2.C2 17:59:39
1.TW4 17:59:39
2.TW1
2.TW2
2.TW3 17:59:43 (with command=None)

Basically what is wrong is that when 2.C1 happened the task status should have been QUEUED (and was!), but it was relabeled to KILLED (actually I could not find log evidence for this, this log part is confusing, but it is the only explanation I could find for what happened)

Then 1.TW4 happened (and before 2.TW2 that gets the second kill data), and set the command to None causing the issue.

I believe that in the context of the status rewriting we should take this into account, and avoid relabeling of DB states..

[1]
grep 160515_130804:duchardt_crab_TTJets_SemiLeptDecays-PU_S10_SP /build/srv-logs/vocms*/crabserver/crabserver-20160515.log

[15/May/2016:17:58:55]  RESTSQL:zaLtlnpdFcMe
[15/May/2016:17:59:08]  RESTSQL:wAaytkEuSqDm
[15/May/2016:17:59:28]  RESTSQL:cDOSsKbYdWLR
[15/May/2016:17:59:28]  RESTSQL:cDOSsKbYdWLR
[15/May/2016:17:59:39]  RESTSQL:zUmgwjLOXhBH
[15/May/2016:17:59:39]  RESTSQL:cDOSsKbYdWLR
[15/May/2016:17:59:39]  RESTSQL:cDOSsKbYdWLR
[15/May/2016:17:59:43]  RESTSQL:BAMEHudJePsp

**RESTSQL:zaLtlnpdFcMe** 
...
[15/May/2016:17:58:55]  RESTSQL:zaLtlnpdFcMe executemany: ([{'status': 'NEW', 'taskname': '160515_130804:duchardt_crab_TTJets_SemiLeptDecays-PU_S10_SP', 'command': 'KILL', 'warnings': '[]'}],) {}
...
The user first crab kill

**RESTSQL:wAaytkEuSqDm** 
...
[15/May/2016:17:59:08]  RESTSQL:wAaytkEuSqDm executemany: ([{'status': 'QUEUED', 'taskname': '160515_130804:duchardt_crab_TTJets_SemiLeptDecays-PU_S10_SP', 'command': 'KILL'}],) {}
...
Task Worker put it to QUEUED (where is HOLDING?)

**RESTSQL:zUmgwjLOXhBH** 1.TW4
...
[15/May/2016:17:59:39]  RESTSQL:zUmgwjLOXhBH executemany: ([{'status': 'KILLED', 'taskname': '160515_130804:duchardt_crab_TTJets_SemiLeptDecays-PU_S10_SP', 'command': None}],) {}
...
TW finishes to kill (Notice 'command' is None!)

**RESTSQL:cDOSsKbYdWLR** 
...
[15/May/2016:17:59:39]  RESTSQL:cDOSsKbYdWLR executemany: ([{'status': 'NEW', 'taskname': '160515_130804:duchardt_crab_TTJets_SemiLeptDecays-PU_S10_SP', 'command': 'KILL', 'warnings': '[]'}],) {}
...
Another crab kill

**RESTSQL:BAMEHudJePsp**
...
[15/May/2016:17:59:43]  RESTSQL:BAMEHudJePsp executemany: ([{'status': 'QUEUED', 'taskname': '160515_130804:duchardt_crab_TTJets_SemiLeptDecays-PU_S10_SP', 'command': 'None'}],) {}
...
TW acquires it
